File: apps/companion/native/android/src/main/kotlin/to/kala/reach/companion/mobile/SigningGate.kt
package to.kala.reach.companion.mobile

/**
 * What the application may sign, and when.
 *
 * The same rule as on the other platform: the application signs an action only after its required
 * authentication, and a receiver or a background worker never signs at all. A worker holds one key
 * that opens a preview and can do nothing else.
 */

/** Whether this process has the authentication signing requires. */
sealed interface AuthenticationState {
    object None : AuthenticationState

    data class Verified(val atMillis: Long) : AuthenticationState
}

/** Why a signature was refused. */
enum class SigningRefusal {
    NOT_AUTHENTICATED,
    AUTHENTICATION_EXPIRED,
    NOT_THE_APPLICATION
}

/** How long an authentication stands before the application asks again. */
const val AUTHENTICATION_LIFETIME_MILLIS: Long = 5 * 60 * 1000

/** Decides whether an action may be signed now. */
data class SigningGate(
    /** True in the application's own process, false in a receiver or a worker. */
    val isApplication: Boolean,
    val state: AuthenticationState
) {
    /** The refusal, or null when the action may be signed. */
    fun refusal(nowMillis: Long): SigningRefusal? {
        if (!isApplication) return SigningRefusal.NOT_THE_APPLICATION
        return when (state) {
            is AuthenticationState.None -> SigningRefusal.NOT_AUTHENTICATED
            is AuthenticationState.Verified ->
                when {
                    nowMillis < state.atMillis -> SigningRefusal.AUTHENTICATION_EXPIRED
                    nowMillis - state.atMillis > AUTHENTICATION_LIFETIME_MILLIS ->
                        SigningRefusal.AUTHENTICATION_EXPIRED
                    else -> null
                }
        }
    }
}

File: apps/companion/native/ios/KalaReachNative/SigningGate.swift
//
//  What the main application may sign, and when.
//
//  Section 16 puts one rule on the application half: it signs an action only after its required
//  authentication. The extension never signs anything at all, and it holds a key that cannot sign:
//  a preview key opens a preview and does nothing else.
//
//  The gate is a value rather than a flag on a singleton, so the rule is a thing a test can hold
//  and a caller cannot forget to ask.
//

import Foundation

/// Whether this process has the authentication signing requires.
enum AuthenticationState: Equatable {
    /// Nothing has been proved in this run.
    case none
    /// The person proved themselves, at a moment.
    case verified(atMilliseconds: UInt64)
}

/// Why a signature was refused.
enum SigningRefusal: Error, Equatable {
    /// Nothing has been proved in this run.
    case notAuthenticated
    /// The proof is older than this application accepts.
    case authenticationExpired
    /// The extension asked, and an extension never signs.
    case notTheApplication
}

/// How long an authentication stands before the application asks again.
let authenticationLifetimeMilliseconds: UInt64 = 5 * 60 * 1000

/// Decides whether an action may be signed now.
struct SigningGate {
    /// True in the application, false in every extension.
    let isMainApplication: Bool
    let state: AuthenticationState

    /// Admits an action for signing, or says why it is refused.
    func admit(nowMilliseconds: UInt64) throws {
        guard isMainApplication else { throw SigningRefusal.notTheApplication }
        switch state {
        case .none:
            throw SigningRefusal.notAuthenticated
        case let .verified(at):
            guard nowMilliseconds >= at else { throw SigningRefusal.authenticationExpired }
            guard nowMilliseconds - at <= authenticationLifetimeMilliseconds else {
                throw SigningRefusal.authenticationExpired
            }
        }
    }
}
